### PR TITLE
Obstacle Queries

### PR DESCRIPTION
**Pull request by kripken**
#6897 solves a fuzz bug in a not very clean way (see discussion near the end). After thinking a lot about it I think the reason we can't find a non-awkward way to fix it is that we are missing some kind of basic capability in the optimizer, and this PR is a proposal for an addition to help there.

Details of the new API appear in `local-graph.h`, specifically `getSetInfluencesGivenObstacle()`. Briefly, the idea is that we want to figure out if moving a `local.set` to another position would change observable behavior or not. One way to do that is to construct a LocalGraph, find what gets it affects, then (temporarily) modify the IR to the state after the optimization and once more construct a LocalGraph and compute the gets. Then if the gets are unchanged, the movement is safe. This works, but is slow. Instead, the proposed API here lets us use a single LocalGraph for this purpose, by

1. Specifying in advance the set of places we can move things to, and
2. Conduct individual queries on those places, as a very cheap operation.

Specifically, we can ask "what gets do I reach, if this location would block the flow?" E.g.
```wat
(local.set $x (A))
(call $foo)
(local.get $x)
(nop)
```
If that `call` is an obstacle then this set has no more gets. If the `nop` is an obstacle then nothing changes. Each call to `getSetInfluencesGivenObstacle()` specifies the current obstacle, which lets us learn things about the impacts of code motion. For example, in the above it would be safe to move the `local.set` to the `call`, but not the `nop`.

This makes #6897 trivial: we can query to see if moving the `local.set` after the `struct.set` is safe or not. I believe this PR could be used in more places as well, e.g. MergeLocals constructs two LocalGraphs wastefully.

Is there perhaps an existing name for this in compilers, or a more standard concept we could use?

### CODE REVIEW

**Comment by kripken** [comment]
(This depends on #7038 and builds fail without it.)

File: src/ir/LocalGraph.cpp [hunk 1]
```diff
@@ -302,12 +305,8 @@ struct LocalGraphFlower
   // initially, but instead fill in these data structures that let us do so
   // later for individual gets. Specifically we need to find the location of a
   // local.get in the CFG.
-  struct BlockLocation {
-    // The basic block an item is in.
-    FlowBlock* block = nullptr;
-    // The index in that block that the item is at.
-    Index index;
-  };
+  using BlockLocation = std::pair<FlowBlock*, Index>;
```
Comment (tlively) [inline src/ir/LocalGraph.cpp:308]: Why this change? It seems that having the custom field names would be better.

File: src/ir/local-graph.h [hunk 2]
```diff
@@ -178,7 +178,11 @@ struct LocalGraph : public LocalGraphBase {
 struct LocalGraphFlower;
 
 struct LazyLocalGraph : public LocalGraphBase {
-  LazyLocalGraph(Function* func, Module* module = nullptr);
+  // We optionally receive an expression class to consider relevant for obstacle
+  // queries, see below. (Only obstacles of this class can be queried later.)
+  LazyLocalGraph(Function* func,
+                 Module* module = nullptr,
+                 std::optional<Expression::Id> obstacleClass = std::nullopt);
```
Comment (tlively) [inline src/ir/local-graph.h:185]: Taking a single expression ID seems oddly restrictive for a general utility like this. Could we make this a general predicate function instead, perhaps?

File: src/ir/local-graph.h [hunk 3]
```diff
@@ -228,7 +232,51 @@ struct LazyLocalGraph : public LocalGraphBase {
...
+  //
+  // An alternative to such obstacle queries is to just modify BinaryenIR,
+  // construct a new LocalGraph, and see what gets it has (and whether they
+  // match the original ones). Obstacle queries are basically a more efficient
+  // way to do that, by reusing the same LocalGraph for all such queries. To do
+  // so we must decide which things can play the role of obstacle
+  // (obstacleClass), and we then place them in the LocalGraph's internal IR.
+  // A query for a particular obstacle is then easy and cheap.
+  SetInfluences getSetInfluencesGivenObstacle(LocalSet* set,
+                                              Expression* obstacle);
```
Comment (tlively) [inline src/ir/local-graph.h:275]: I think it would be more straightforward to frame this in terms of reachability. Ultimately what we want to know is whether control flow reaches the same set of gets from some set and some arbitrary other instruction. Computing and comparing the set of reached gets is one possible implementation of that, but computing that second set is not ultimately the problem we want to solve, so it seems odd to have an API for it.

Can we have an API like `bool hasSameSetInfluences(LocalSet* set, Expression* expr)`? It would just require moving the comparison of the sets from the client to the utility.

File: src/ir/LocalGraph.cpp [hunk 4]
```diff
@@ -444,6 +440,73 @@ struct LocalGraphFlower
       }
     }
   }
+
+  // Given a bunch of gets, see if any of them reach the given set despite the
+  // obstacle expression stopping the flow whenever it is reached.
+  LocalGraphBase::SetInfluences
+  getSetInfluencesGivenObstacle(const LocalGraphBase::SetInfluences& gets,
```
Comment (tlively) [inline src/ir/LocalGraph.cpp:447]: This all seems very similar to code that probably already exists. Can we use a bool template parameter and `if constexpr` to have a single implementation that can be configured to consider obstacles or not?

File: src/ir/local-graph.h (hunk 3 above)
Comment (kripken) [reply src/ir/local-graph.h:275]: `hasSameSetInfluences(set, expr)` would need to either

1. Start from `expr` and flow forward to find the gets, which is usually less efficient (it will reach dead ranges at the end of the function, past any gets), or
2. Start from all gets of the relevant local index and flow backwards. This is less efficient than starting from a known set of gets, which is what the current PR does.

Overall the problem we have here is not fully general: we want to consider a small "tweak" to a `local.set`, not an arbitrary new position for it.

But I do see what you're saying that the API here is a little odd. Thinking now, maybe another option is `findGetsReachingExpr(gets, expr)` which would flow back from those gets specifically. It would be called using `findGetsReachingExpr(getSetInfluences(set), expr)` (that is, the gets are the original gets that the set reaches), but might have more general uses later too. This should be efficient as it flows back from the relevant gets, and it will stop at either `expr` or the `set` (we don't need to tell it the `set` explicitly, since it will stop at any set of that index). Thoughts?

Comment (tlively) [reply src/ir/local-graph.h:275]: Hmm, maybe I'm misunderstanding something. My hope was that we could just inline some of the current client code into the utility to provide a simpler interface where the client just does one call. I wasn't thinking that the actual algorithm would have to change.

Comment (kripken) [reply src/ir/local-graph.h:275]: Anything else I can clarify here?

File: src/ir/LocalGraph.cpp (hunk 1 above)
Comment (kripken) [reply src/ir/LocalGraph.cpp:308]: The names are nice, but in practice this struct is always immediately destructured,
```cpp
auto [block, index] = ..
```
so we are not losing names here. And, using a `std::pair` gives us hashing for free without writing new boilerplate, which is needed in this PR (it uses a map on this type).

File: src/ir/local-graph.h (hunk 3 above)
Comment (kripken) [reply src/ir/local-graph.h:275]: In discussion offline we agreed that an API like `findGetsReachedBy(gets, expr)` would make sense. However, mid-way into writing the code to switch to it I realized that doesn't quite work. Consider this:

```wat
(block $label
 ;; A struct.set that may be skipped by the br in the if.
 (struct.set $struct 0  ;; structSet
  (local.tee $struct  ;; localTee
   (struct.new_default $struct)
  )
  (if (result i32)
   (..)
   (then
    (br $label)
   )
   (else
    (i32.const 0)
   )
  )
 )
)
(drop
 (struct.get $struct 0
  (local.get $struct)  ;; get
 )
)
```
If we ask `findGetsReachedBy({ get }, structSet)` then we get `{ get }`  - the structSet reaches that get. But the br may skip over it! So while the structSet and the localTee can both reach the get, one of them might not be passed by on the way. That is, we can go straight from the localTee to the get, skipping the structSet.

And that is the difference we care about here. If the localTee were moved to the structSet, and then skipped, then the get would read a different value from the local (whatever was in the local before all this code).

To compute this "must pass by X on the way from A to B" we do need both the gets, and the set, and the obstacle along the way, I think. So an "obstacle query" is still the best idea I can think of.

Comment (tlively) [reply src/ir/local-graph.h:275]: Can we do `findGetsDominatedByExpr`?

Comment (kripken) [reply src/ir/local-graph.h:275]: Good idea, yeah, that might work. "Dominated" here meaning "is the only set for them?" (so it dominates in the CFG sense, and also no other set is between them). I'll write that up.

Comment (kripken) [reply src/ir/local-graph.h]: After more offline discussion we agreed to rename the API to "canMoveSet", which goes directly to the point of what we are computing, keeping the API itself simple and easy to explain. The implementation does do an obstacle analysis to compute things, but that is an internal detail.

File: src/ir/local-graph.h (hunk 2 above)
Comment (kripken) [reply src/ir/local-graph.h]: A predicate function is possible, but would be slower I think.

I think this API is not odd, though: it is reasonable to consider a particular expression. For example, in HeapStoreOptimizations we always want to move LocalSets to StructSets. Each optimization has a pattern it optimizes, and a particular expression type that we move to. We may want to generalize this more later, of course.

File: src/ir/local-graph.h [hunk 5]
```diff
@@ -228,7 +232,39 @@ struct LazyLocalGraph : public LocalGraphBase {
...
+  // If we move the set to the call, then the get on line 2 could observe a
+  // different value (the pre-existing value in that local, before line 1), and
+  // we'd return that get here.
+  //
+  // |to| must be of the class |queryClass| that is defined during construction.
+  // That is, we must decide ahead of time which places we want to query about
+  // moving LocalSets to, and can then issue queries on any of those, in an
+  // efficient manner.
+  //
+  // This assumes that |to| is in a position dominated by |set|, that is we are
```
Comment (tlively) [inline src/ir/local-graph.h:257]: Not "dominated," just "reached by," right? The whole point is that there might be control flow that skips the destination location, making the move invalid and also meaning that the destination is not dominated by the source.

**Review by tlively — APPROVED** [review]

File: src/ir/local-graph.h (hunk 5 above)
Comment (kripken) [reply src/ir/local-graph.h:257]: This is the situation we have:
```c
x = C; // set
if (..) goto B;
foo(x); // use
```
We might branch away before reaching the use of `x`, but the use is still dominated by the set (since all paths from the entry to `foo` must go through the set).

Just being reached is not enough:
```c
while (1) {
  foo(x);
  bar(x);
  x = C; // the original set
  if (..) goto B;
}
```
Here the original set reaches `foo(x)` (using the backedge of the loop), but it does not dominate it. Note how `foo(x)`'s influence, were it a set, is larger than `x = C` (`foo` definitely reaches `bar`, while `x = C` might not).

File: src/ir/LocalGraph.cpp [hunk 6]
```diff
@@ -444,6 +440,75 @@ struct LocalGraphFlower
...
+              break;
+            }
+          } else if (action == obstacle) {
+            // We ran into the obstacle. Halt this flow.
+            break;
+          }
+          // TODO: If the action is one of the gets we are scanning, then
+          // either we have processed it already, or will do so later, and we
+          // can halt. As an optimization, we could check if we've processed
+          // it already and act accordingly.
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/ir/LocalGraph.cpp:505]: Hmm, that is a very good point... in general backwards is faster, but here we expect the obstacle to stop us in almost all cases. I'll experiment with that.

[the remark this answers is not in the document]
Comment (kripken) [reply src/ir/LocalGraph.cpp:516]: Ah, thinking on this some more, I still prefer backwards. I added a comment why, but basically, the original situation that led to all this, calls in `struct.set` values, each get a branch away. If we follow one such branch we may end up scanning lots of irrelevant code, while gets tend to be close to the set.

[the remark this answers is not in the document]
Comment (tlively) [reply src/ir/LocalGraph.cpp:516]: In principle searching backward can also end up scanning all the prior blocks in the function, right? So either direction can pretty easily end up doing way too much work.

File: src/ir/local-graph.h (hunk 5 above)
Comment (tlively) [reply src/ir/local-graph.h:257]: Right, I got confused again. Thanks!

**Review by tlively — APPROVED** [review]

File: src/ir/LocalGraph.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (kripken) [reply src/ir/LocalGraph.cpp:516]: True, but it isn't symmetrical. Going backwards tends to reach a dominating set which is close by. Only more rarely does going backwards needs to get all the way to the function entry (to use a param value or the default value). And in the code here, we are considering moving a set, so we definitely have an actual set.

File: src/ir/local-graph.h (hunk 5 above)
Comment (kripken) [reply src/ir/local-graph.h:257]: This is tricky stuff... I keep needing to re-read the sample code.
